### DISCUSSION TITLE
tutorials/scripting/singletons_autoload

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#45] [User-contributed notes]
# tutorials/scripting/singletons_autoload

Introduction: Godot's scene system, while powerful and flexible, has a drawback: there is no method for storing information (e.g. a player's score or inventory) that is needed by more than one scen...

https://docs.godotengine.org/en/stable/tutorials/scripting/singletons_autoload.html

**Comment by vbettaque** [comment]
You can simply write
```
current_scene = root.get_child(-1)
```
instead of 
```
current_scene = root.get_child(root.get_child_count() - 1)
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10152

**Reply by ntlblpm** [reply]
You can also simply write `current_scene = get_tree().current_scene` for the entire `_ready()` function

**Comment by yemibox51** [comment]
Be wary of using this implementation of a scene switcher and then later calling `get_tree().reload_current_scene()`. `reload_current_scene()` replaces the `current_scene` with a new object and so the previous one will be destroyed and be set to `null`. 

To fix this, just make sure you are grabbing the new `current_scene` shortly after calling `reload_current_scene()` or just make sure to set `current_scene` to `get_tree().current_scene` before calling `_deferred_goto_scene.call_deferred(path)` like so:

```gdscript
func goto_scene(path):
    current_scene = get_tree().current_scene
	_deferred_goto_scene.call_deferred(path)
```

**Comment by RudyFisher7** [comment]
Autoloads are neat. Note also that since the introduction of static variables, a C#-esc singleton pattern can also be used.

```gdscript
class_name InteractableService3D
extends Node


static var _singleton: InteractableService3D = null
static var singleton: InteractableService3D:
	get:
		return _singleton

func _init() -> void:
	if singleton == null:
		_singleton = self
```

**Reply by RudyFisher7** [reply]
Just in case someone might find this helpful. Some benefits/trade-offs of using this approach vs. autoloads:

1. The singleton can be part of the scene you transition to and so will be cleaned up when transitioned away from it (say to the main menu).

```gdscript
func _enter_tree() -> void:
	if singleton == null:
		_singleton = self


func _exit_tree() -> void:
	if singleton == self:
		_singleton = null
```
2. You can add child nodes to the singleton using the scene tree inspector while autoloads require adding them via code.

**Reply by RudyFisher7** [reply]
(I'm getting too geeked out about this...)

Here is another benefit to using traditional-style singletons:

You can call a static function on your singleton node's class_name like so below. Since this is a static function, there is no need to add an autoload.
```gdscript
func _enter_tree() -> void:
	var input := PlayerInput3D.from_device_id(0)
	var icon := InputService.get_icon(input.face_a, input)
	%InteractionMenu.prompt_icon = icon
	%InteractionMenu.prompt_icon_filter = CanvasItem.TextureFilter.TEXTURE_FILTER_NEAREST
```

You can have fallback functionality if the singleton hasn't been set to a valid instance (like upon entering the scene tree).
```gdscript
static func get_icon(action: StringName, input: PlayerInput3D) -> Texture2D:
	var result: Texture2D = null
	
	var path: String = ""
	if singleton:
		path = singleton.controls_icons_root_folder.path_join("steam_deck")
	else:
		path = "res://lib/common/input/assets/controls/default"
	
	match action:
		input.face_a:
			path = path.path_join("a.png")
			result = ResourceLoader.load(path, "Texture2D") as Texture2D
		_:
			pass
	
	return result
```

Comparing this to the autoload, an instance MUST exist (set in project settings) otherwise the first snippet would cause a parse error and your project won't be able to run. 

This opens many opportunities, say you inherit from InputService as SteamInputService. You can safely swap out concrete implementations this way using a dependency injector without your project breaking.

**Reply by Nadav-Ruskin** [reply]
Here's how I set up gdextension singletons (and a [signal bus pattern](https://github.com/godotengine/godot-docs-user-notes/discussions/5#discussioncomment-8124099))

Credit to [xana43](https://github.com/xana43) for providing an example [here](https://github.com/godotengine/godot-cpp/issues/1623#issuecomment-2444498863)

```cpp
//  register_type.cpp
#include <godot_cpp/classes/engine.hpp>
#include "signal_bus.hpp"

using namespace godot;

static SignalBus *SignalBus_singleton = nullptr;

void initialize_gdextension_types(ModuleInitializationLevel p_level)
{
    if (p_level != MODULE_INITIALIZATION_LEVEL_SCENE) {
        return;
    }
    
    GDREGISTER_CLASS(SignalBus);
    SignalBus_singleton = memnew(SignalBus);
    Engine::get_singleton()->register_singleton("SignalBus", SignalBus::get_singleton());
}

void uninitialize_gdextension_types(ModuleInitializationLevel p_level) {
    if (p_level != MODULE_INITIALIZATION_LEVEL_SCENE) {
        return;
    }

    Engine::get_singleton()->unregister_singleton("SignalBus");
    memdelete(SignalBus_singleton);
    SignalBus_singleton = nullptr;
}

// normal extern C code below...
```
```cpp
// signal_bus.hpp
#pragma once
#include <godot_cpp/classes/node.hpp>


using namespace godot;


// class Prompt: public Sprite2D{

class SignalBus final: public Object {
    private:
        GDCLASS(SignalBus, Object);
        static SignalBus *singleton;

    protected:
        static void _bind_methods();
    
    public:
        SignalBus();
        ~SignalBus();
        static SignalBus *get_singleton();

        Signal points_gained{this,"points_gained"};
};
```
```cpp
// signal_bus.cpp
#include "signal_bus.hpp"
#include <godot_cpp/variant/signal.hpp>


// protected:
void SignalBus::_bind_methods(){

    ADD_SIGNAL(MethodInfo("points_gained", PropertyInfo(Variant::INT, "value")));
}

SignalBus *SignalBus::singleton = nullptr;

SignalBus *SignalBus::get_singleton() {
    return singleton;
}

SignalBus::SignalBus(){
    ERR_FAIL_COND(singleton != nullptr);
    singleton = this;
}

SignalBus::~SignalBus() {
    ERR_FAIL_COND(singleton != this);
    singleton = nullptr;
}
```
```cpp
// emitter.cpp
void Emitter::emit_signal() {    
    SignalBus::get_singleton()->points_gained.emit(1);
}
```
```cpp
// receiver.cpp
void Receiver::_ready(){
    SignalBus::get_singleton()->points_gained.connect(callable_mp(this, &Receiver::points_gained));
}
void Receiver::points_gained(int number_of_points){
    order_failed();
    godot::UtilityFunctions::print("got " + number_of_points + " points");
}
```

**Comment by cgudrian** [comment]
Why does `global.gd` manage its own local copy of `current_scene` instead of using `get_tree().current_scene` all the way through?

**Reply by PatchworkGoose** [reply]
Late but answering in case anyone else would like to know. 

If we called ```get_tree().current_scene``` after ```current_scene.free()``` it would return ```<Object#null>``` because the current scene has been freed and the new scene is not yet initialized. 

This is the problem being solved by using an Autoload. In this case we are using ```global.gd``` to hold ```var current_scene``` in memory allowing us to know exactly which node is active and access it immediately.

If you want to  see this behavior play out, use something like ```print("Global var: ", current_scene, ". Get Tree: ", get_tree().current_scene)``` before each line in the function and at the end.

**Comment by Bobrokus** [comment]
[There's another way to define auto-loads in C#](https://www.jetbrains.com/guide/gamedev/tutorials/singletons-autoloads-godot-csharp/)

**Comment by DeletedBunny** [comment]
I would like to add a workaround for an issue I've been having with the C# way of accessing autoloads. 

I implemented my autoload in the form with the `Instance` property as in the example, however I was needed to access my class through the namespace name for it to compile. In my case it was `EventBus.EventBus.Instance` otherwise it wouldn't compile. Apparently the namespace being the same name as the autoload causes it to be ambiguous when trying to access. After changing my folder name and namespace to something else now I am able to access it like in the example `EventBus.Instance`.

**Comment by UrbanNuke** [comment]
How to prevent replacing res path to uid in autoload godot.project by Godot itself?
We have some autoloads from addons folder which installing by package manager, and every contributor has different uids, looks like a problem
